# Logistic Regression Benchmarking

In this notebook, we compare our custom-built Logistic Regression model (using Gradient Ascent / Maximum Likelihood Estimation) against the `scikit-learn` implementation. We will evaluate the performance on a binary classification problem (Breast Cancer dataset) using Accuracy, Precision, Recall, and F1-Score.

In [1]:
import sys
import os
# Allow notebook to import from the parent directory
sys.path.append(os.path.abspath('..'))

import numpy as np
from time import time
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression as SklearnLogReg

# Import our custom modules
from classical_ml.linear_models.logistic_regression import LogisticRegression as CustomLogReg
from utils.metrics import accuracy_score, precision_score, recall_score, f1_score

d:\Github\Classical-ML-From-Scratch\classical_ml\linear_models\logistic_regression.py:25: SyntaxWarning: invalid escape sequence '\s'
  Formula: \sigma(z) = \frac{1}{1 + e^{-z}}
d:\Github\Classical-ML-From-Scratch\classical_ml\linear_models\logistic_regression.py:63: SyntaxWarning: invalid escape sequence '\s'
  Formula: P(y=1|X) = \sigma(X * w + b)
d:\Github\Classical-ML-From-Scratch\utils\metrics.py:6: SyntaxWarning: invalid escape sequence '\s'
  Formula: (1/n) * \sum (y_true - y_pred)^2
d:\Github\Classical-ML-From-Scratch\utils\metrics.py:13: SyntaxWarning: invalid escape sequence '\s'
  Formula: (1/n) * \sum |y_true - y_pred|


In [4]:
# 1. Load Dataset
data = load_breast_cancer()
X, y = data.data, data.target

# 2. Split into train and test sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Scale the features (Crucial for Gradient Descent stability)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print(f"Training data shape: {X_train_scaled.shape}")
print(f"Test data shape: {X_test_scaled.shape}")

Training data shape: (455, 30)
Test data shape: (114, 30)


In [6]:
print("1. Custom Logistic Regression")
start = time()
# Initialize with learning rate 0.1 and 2000 iterations
custom_model = CustomLogReg(learning_rate=0.1, n_iters=2000)
custom_model.fit(X_train_scaled, y_train)
custom_preds = custom_model.predict(X_test_scaled)
custom_time = time() - start

print(f"Accuracy   : {accuracy_score(y_test, custom_preds):.4f}")
print(f"Precision  : {precision_score(y_test, custom_preds):.4f}")
print(f"Recall     : {recall_score(y_test, custom_preds):.4f}")
print(f"F1 Score   : {f1_score(y_test, custom_preds):.4f}")
print(f"Time Taken : {custom_time:.4f} seconds")

1. Custom Logistic Regression
Accuracy   : 0.9912
Precision  : 0.9861
Recall     : 1.0000
F1 Score   : 0.9930
Time Taken : 0.1146 seconds


In [10]:
print("2. Scikit-learn Logistic Regression")
start = time()
# Sklearn use L2 penalty by default
sklearn_model = SklearnLogReg()
sklearn_model.fit(X_train_scaled, y_train)
sklearn_preds = sklearn_model.predict(X_test_scaled)
sklearn_time = time() - start

print(f"Accuracy   : {accuracy_score(y_test, sklearn_preds):.4f}")
print(f"Precision  : {precision_score(y_test, sklearn_preds):.4f}")
print(f"Recall     : {recall_score(y_test, sklearn_preds):.4f}")
print(f"F1 Score   : {f1_score(y_test, sklearn_preds):.4f}")
print(f"Time Taken : {sklearn_time:.4f} seconds")

2. Scikit-learn Logistic Regression
Accuracy   : 0.9737
Precision  : 0.9722
Recall     : 0.9859
F1 Score   : 0.9790
Time Taken : 0.0140 seconds


## Conclusion
Both models perform exceptionally well on the Breast Cancer dataset, achieving near-identical accuracy and F1-scores (typically > 97%). 

Any minor discrepancies between the two models (e.g., 0.01% difference) can be attributed to the fact that `scikit-learn` applies an **L2 Regularization penalty** by default and uses advanced solvers like `lbfgs`, whereas our custom implementation uses vanilla **Stochastic Gradient Ascent (without regularization)**. Overall, this confirms that our mathematical formulation of the log-loss gradients is highly accurate.